aptos-move/framework/aptos-stdlib/sources/pool_u64.move

```diff
@@ -0,0 +1,535 @@
+    public fun create(shareholders_limit: u64): Pool {
+        // Default to a scaling factor of 1 (effectively no scaling).
+        create_with_scaling_factor(shareholders_limit, 1)
+    }
+
+    /// Create a new pool with custom `scaling_factor`.
+    public fun create_with_scaling_factor(shareholders_limit: u64, scaling_factor: u64): Pool {
+        Pool {
+            shareholders_limit,
+            total_coins: 0,
+            total_shares: 0,
+            shares: simple_map::create<address, u64>(),
+            shareholders: vector::empty<address>(),
+            scaling_factor,
+        }
+    }
+
+    /// Destroy an empty pool. This will fail if the pool has any balance of coins.
+    public fun destroy_empty(pool: Pool) {
+        assert!(pool.total_coins == 0, error::invalid_state(EPOOL_IS_NOT_EMPTY));
+        let Pool {
+            shareholders_limit: _,
+            total_coins: _,
+            total_shares: _,
+            shares: _,
+            shareholders: _,
+            scaling_factor: _,
+        } = pool;
+    }
+
+    /// Return `pool`'s total balance of coins.
+    public fun total_coins(pool: &Pool): u64 {
+        pool.total_coins
+    }
+
+    /// Return the total number of shares across all shareholders in `pool`.
+    public fun total_shares(pool: &Pool): u64 {
+        pool.total_shares
+    }
+
+    /// Return true if `shareholder` is in `pool`.
+    public fun contains(pool: &Pool, shareholder: address): bool {
+        simple_map::contains_key(&pool.shares, &shareholder)
+    }
+
+    /// Return the number of shares of `stakeholder` in `pool`.
+    public fun shares(pool: &Pool, shareholder: address): u64 {
+        if (contains(pool, shareholder)) {
+            *simple_map::borrow(&pool.shares, &shareholder)
+        } else {
+            0
+        }
+    }
+
+    /// Return the balance in coins of `shareholder` in `pool.`
+    public fun balance(pool: &Pool, shareholder: address): u64 {
+        let num_shares = shares(pool, shareholder);
+        shares_to_amount(pool, num_shares)
+    }
+
+    /// Return the list of shareholders in `pool`.
+    public fun shareholders(pool: &Pool): vector<address> {
+        pool.shareholders
+    }
+
+    /// Return the number of shareholders in `pool`.
+    public fun shareholders_count(pool: &Pool): u64 {
+        vector::length(&pool.shareholders)
+    }
+
+    /// Update `pool`'s total balance of coins.
+    public fun update_total_coins(pool: &mut Pool, new_total_coins: u64) {
+        pool.total_coins = new_total_coins;
+    }
+
+    /// Allow an existing or new shareholder to add their coins to the pool in exchange for new shares.
+    public fun buy_in(pool: &mut Pool, shareholder: address, coins_amount: u64): u64 {
+        if (coins_amount == 0) return 0;
+
+        let new_shares = amount_to_shares(pool, coins_amount);
+        assert!(MAX_U64 - pool.total_coins >= coins_amount, error::invalid_argument(EPOOL_TOTAL_COINS_OVERFLOW));
```

Comment:
It is nice to have explicit error here. I think VM also throws overflow error as well